ta/pkcs11/src/pkcs11_token.c

```diff
@@ -701,3 +701,118 @@ uint32_t entry_ck_session_info(struct pkcs11_client *client,
 
 	return PKCS11_CKR_OK;
 }
+
+uint32_t entry_ck_token_initialize(uint32_t ptypes, TEE_Param *params)
+{
+        const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+	char label[PKCS11_TOKEN_LABEL_SIZE + 1] = { 0 };
+	struct pkcs11_client *client = NULL;
+	struct pkcs11_session *sess = NULL;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	struct serialargs ctrlargs = { };
+	struct ck_token *token = NULL;
+	TEE_Param *ctrl = params;
+	uint32_t token_id = 0;
+	uint32_t pin_size = 0;
+	void *pin = NULL;
+
+	if (ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get(&ctrlargs, &token_id, sizeof(uint32_t));
```

Comment:
From PKCS11 spec, there are precedence requirement on some (very few) return values.
Invalid slot ID (or token ID, same) is one of them.

You can find below a suggestion to factorize this with a friendly `serialargs_get_session()` helper function:
https://github.com/etienne-lms/optee_os/pull/21/commits/6a1e6d9c7bd12817ea89ab2e7535a14615797d1e

Reply:
`PKCS11_CKR_SLOT_ID_INVALID`, does not take precedence over any other error code (see 5.1.6  All other Cryptoki function return values).
The precedence rule (as I understand it) only is applicable if there's multiple competing errors. So if unpacking of arguments fail, we return `PKCS11_CKR_ARGUMENTS_BAD`, there's no other competing error code at this stage.

I'll move and update the `pin_size` check below though. I guess it makes sense to check the arguments in the order they were supplied to `C_InitToken()`.

Reply:
You're right, sorry. I mixed up with the session handle that should take precedence (CKR_SESSION_HANDLE_INVALID).

I think my comment could apply to the commits where session_handle is read from the client arguments.